CNF-6946: feat: expand support to render based on files in dir

eggfoobar opened the pull request:
This PR updates our render command to be more inline with a more flexible approach during the bootstrapping process.

Typically operators that need to generate files during bootstrapping might need some initial cluster state information in order to correctly render the resource files, some have static information that is retrieved and others more dynamic. In this approach we take the pattern used by MCO where the manifest directory is inspected and the relevant files are read in and parsed to render out the resulting configurations. ([see](https://github.com/openshift/machine-config-operator/blob/bff62e7ea81a3180860b28068fd33a7209475f14/pkg/controller/bootstrap/bootstrap.go#L52-L232))

This affords us some niceties like not being over prescriptive with file names for performance profile manifests that are supplied during bootstrap, as well as the ability to get the bootstrap cluster state and inform our rendering of the configuration files. 

Signed-off-by: ehila <ehila@redhat.com>

eggfoobar commented:
/retest-required

eggfoobar commented:
/cc @MarSik 
/assign @yanirq
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

yanirq commented:
The render tests should be aligned to use the new render command  https://github.com/openshift/cluster-node-tuning-operator/blob/e43b1372c35826f65137978c51a0836994a9a2c9/test/e2e/performanceprofile/functests-render-command/1_render_command/render_test.go#L81

eggfoobar commented:
> The render tests should be aligned to use the new render command
> 
> https://github.com/openshift/cluster-node-tuning-operator/blob/e43b1372c35826f65137978c51a0836994a9a2c9/test/e2e/performanceprofile/functests-render-command/1_render_command/render_test.go#L81

Didn't see that there, updated, thanks! 

Also wanted to bring something to your attention, it seems we updated our runtime class struct to `node.k8s.io/v1` api, but we were still rendering the `node.k8s.io/v1beta1` in my testing that was causing errors when applying it during the bootstrapping process, so I updated it to `node.k8s.io/v1`. In current deployments it seems that gets updated to `node.k8s.io/v1` anyway, so this should be an okay change. What do you think?

https://github.com/openshift/cluster-node-tuning-operator/pull/522/files#diff-8fc0682a821429188c846138f0c000cf3ca68ac1807612f1b6306b33b251b26dR17

yanirq commented:
> > The render tests should be aligned to use the new render command
> > https://github.com/openshift/cluster-node-tuning-operator/blob/e43b1372c35826f65137978c51a0836994a9a2c9/test/e2e/performanceprofile/functests-render-command/1_render_command/render_test.go#L81
> 
> Didn't see that there, updated, thanks!
> 
> Also wanted to bring something to your attention, it seems we updated our runtime class struct to `node.k8s.io/v1` api, but we were still rendering the `node.k8s.io/v1beta1` in my testing that was causing errors when applying it during the bootstrapping process, so I updated it to `node.k8s.io/v1`. In current deployments it seems that gets updated to `node.k8s.io/v1` anyway, so this should be an okay change. What do you think?
> 
> https://github.com/openshift/cluster-node-tuning-operator/pull/522/files#diff-8fc0682a821429188c846138f0c000cf3ca68ac1807612f1b6306b33b251b26dR17

Actually we should not have any occurrences of [node.k8s.io/v1beta1](http://node.k8s.io/v1beta1) and this requires a re scan of our current code to track any additional rouge appearances of this deprecated version.
Thanks for bringing that up.

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

yanirq commented:
@eggfoobar  should we put this PR on hold until https://github.com/openshift/enhancements/pull/1213 gets a final approval ?

eggfoobar commented:
I think this should be fine on it's own, it's needed by the EP but it was already something that we wanted NTO to do during bootstrap. This code doesn't do anything with infrastructure or workload partitioning yet, I have that part of the #431 this is just refining our code to run during bootstrap. @MarSik What do you think?

MarSik commented:
> I think this should be fine on it's own, it's needed by the EP but it was already something that we wanted NTO to do during bootstrap. This code doesn't do anything with infrastructure or workload partitioning yet, I have that part of the #431 this is just refining our code to run during bootstrap. @MarSik What do you think?

I think merging this in isolation is fine as long as the commit message provides good enough rationale. Like pointing out to a common pattern everyone else uses. The directory in - directory out model was the indended one anyway, so I am ok with treating this as a standalone change not coupled to the enhancement.

eggfoobar commented:
Hey @yanirq, I went ahead and updated the commit messages and PR message to provide clear rational on this change.

yanirq commented:
> Hey @yanirq, I went ahead and updated the commit messages and PR message to provide clear rational on this change.

Thanks @eggfoobar.  Before we get to a final round of reviews and merging the PR it would be good to squash the commits first if possible.

jlojosnegros commented:
/lgtm

eggfoobar commented:
Sure thing @yanirq, should be all set.

yanirq commented:
/lgtm
/approve
